Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -21,981 +21,660 @@
 }
 
 /**
  @Status Interoperable
 */
-CGColorSpaceRef CGImageGetColorSpace(CGImageRef img) {
-    // TODO: Consider caching colorspaceRef in CGImageRef
-    CGColorSpaceRef ret = (CGColorSpaceRef) new __CGColorSpace(img->Backing()->ColorSpaceModel());
-
-    return ret;
+bool CGImageGetShouldInterpolate(CGImageRef image) {
+  RETURN_FALSE_IF(!image);
+  return image->Interpolate();
 }
 
 /**
  @Status Interoperable
 */
-size_t CGImageGetBitsPerPixel(CGImageRef img) {
-    if (!img) {
-        TraceWarning(TAG, L"CGImageGetBitsPerPixel: nil!");
-        return 0;
-    }
-
-    return (img->Backing()->BytesPerPixel() << 3);
+bool CGImageIsMask(CGImageRef image) {
+  RETURN_FALSE_IF(!image);
+  return image->IsMask();
 }
 
 /**
  @Status Interoperable
 */
-size_t CGImageGetBitsPerComponent(CGImageRef img) {
-    if (!img) {
-        TraceWarning(TAG, L"CGImageGetBitsPerComponent: nil!");
-        return 0;
-    }
-
-    return (img->Backing()->BitsPerComponent());
+CGColorSpaceRef CGImageGetColorSpace(CGImageRef img) {
+  RETURN_NULL_IF(!img);
+  return img->ColorSpace();
 }
 
 /**
  @Status Interoperable
 */
 size_t CGImageGetWidth(CGImageRef img) {
-    if (!img)
-        return 0;
-
-    return img->Backing()->Width();
+  RETURN_RESULT_IF_NULL(img, 0);
+  return img->Width();
 }
 
 /**
  @Status Interoperable
 */
 size_t CGImageGetHeight(CGImageRef img) {
-    if (!img)
-        return 0;
-
-    return img->Backing()->Height();
+  RETURN_RESULT_IF_NULL(img, 0);
+  return img->Height();
 }
 
 /**
  @Status Interoperable
 */
 void CGImageRelease(CGImageRef img) {
-    CFRelease((id)img);
+  RETURN_IF(!img);
+  CFRelease((CFTypeRef)img);
 }
 
 /**
  @Status Interoperable
 */
 CGImageRef CGImageRetain(CGImageRef img) {
-    CFRetain((id)img);
-    return img;
+  RETURN_NULL_IF(!img);
+  CFRetain((CFTypeRef)img);
+  return img;
 }
 
 /**
  @Status Interoperable
 */
 CGBitmapInfo CGImageGetBitmapInfo(CGImageRef img) {
-    int ret;
-
-    if (img) {
-        ret = img->Backing()->BitmapInfo();
-    } else {
-        TraceWarning(TAG, L"CGImageGetBitmapInfo: Null CGImageRef!");
-        ret = 0;
-    }
-
-    return ret;
+  RETURN_RESULT_IF_NULL(img, 0);
+  return img->BitmapInfo();
 }
 
 /**
- @Status Interoperable
+ @Status Caveat
+ @Notes decode parameter not supported and must be nullptr.
 */
-size_t CGImageGetBytesPerRow(CGImageRef img) {
-    if (!img)
-        return 0;
+CGImageRef CGImageCreateWithJPEGDataProvider(CGDataProviderRef source,
+                                             const CGFloat decode[],
+                                             bool shouldInterpolate,
+                                             CGColorRenderingIntent intent) {
+  RETURN_NULL_IF((source == nullptr) ||
+                 ![(NSObject *)source isKindOfClass:[NSData class]]);
 
-    DWORD ret = 0;
+  NSData *sourceData = (__bridge NSData *)source;
+  CGImageRef imageRef =
+      _CGImageLoadJPEG((void *)[sourceData bytes], [sourceData length]);
```

Comment:
given data could be huge, i'd rather not introduce copies.

---
In reply to: [84143048](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r84143048) [](ancestors = 84143048)